backend/canisters/registry/impl/src/model/message_filters.rs

```diff
@@ -0,0 +1,72 @@
+use registry_canister::MessageFilterSummary;
+use serde::{Deserialize, Serialize};
+use types::TimestampMillis;
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct MessageFilters {
+    last_updated: TimestampMillis,
+    filters: Vec<Filter>,
+}
+
+#[derive(Serialize, Deserialize, Default)]
+struct Filter {
+    regex: Option<String>,
+    last_updated: TimestampMillis,
+}
+
+impl MessageFilters {
+    pub fn add(&mut self, regex: String, now: TimestampMillis) -> Option<u64> {
+        let regex = Some(regex);
+        if self.filters.iter().any(|f| f.regex == regex) {
+            None
+        } else {
+            let id = (self.filters.len() - 1) as u64;
+            self.filters.push(Filter {
+                regex,
+                last_updated: now,
+            });
+            self.last_updated = now;
+            Some(id)
+        }
+    }
+
+    pub fn remove(&mut self, index: u64, now: TimestampMillis) -> bool {
+        let index = index as usize;
+        if index < self.filters.len() {
+            let filter = &mut self.filters[index];
+            filter.regex = None;
```

Comment:
Seems a bit strange to remove the text but leave the Id.
Maybe use a map to store them and hold the latest index separately?
Or set a flag on the entry to mark it removed?

Reply:
We could have `added: Timestamp, removed: Option<Timestamp>`